### ISSUE TITLE
Expose the VENV to child processes

### ISSUE THREAD

**Issue by simaoafonso-pwt** [sio/Makefile.venv#12]
After using this in several projects (thanks for sharing this!), I found myself needing project-specific shell scripts for some tasks.

I wanted to re-use the venv created by this tool, since it has all the necessary tools already installed, but this is not easy to do, since I'm writing a shell script, not Python code. After some trial and error, I found this to work very well:

## Expose the VENV location to the child processes

On the `Makefile`, use:

```make
export VENV
```

## Scripts

On the scripts themselves, I use this snippet on the top:

```sh
#!/bin/bash
if [ -n "$VENV" ]; then
	echo "# Augment PATH with '$VENV'" >&2
	PATH="$(realpath "$VENV"):$PATH"
	export PATH
fi
```
That last `export` re-exports the PATH to child processes, so that they can also use the venv tools. These scripts also work inside the shell from `make bash`, with duplicated PATH.

Another (worse) alternative is to adjust all calls, using something like this:

```sh
"${VENV+$VENV/}python" --version
```

## Makefile Targets

That way you can define target in the "usual" way in the `Makefile`, and all the venv tools are available on the PATH:

```make
output.txt: input.txt | $(VENV)
    location/script <$< >$@
```

---

This requires changes on the scripts themselves, but I think this is the cleanest way.

Just putting this out here in case anyone runs into the same problem.

**Comment by sio** [comment]
Thank you for sharing your workflow!

You could also move the PATH modification snippet into Makefile, that way the scripts would not need to be made aware of $VENV:

```make
export PATH:=$(abspath $(VENV)):$(PATH)
output.txt: input.txt | venv
    location/script <$< >$@
```

Also, be aware that using `$(VENV)` as a dependency in recipes is not the same as using `venv`, and is not recommended. You lose auto-updating capabilities that way - if your requirements.txt changes `$(VENV)` dependency will not trigger an update of the existing virtual environment but `venv` will.

**Comment by simaoafonso-pwt** [comment]
> You could also move the PATH modification snippet into Makefile, that way the scripts would not need to be made aware of $VENV:
> 
> ```makefile
> export PATH:=$(abspath $(VENV)):$(PATH)
> output.txt: input.txt | venv
>     location/script <$< >$@
> ```

Interesting, I tried sillier versions of this envolving `env`, but this seems much cleaner. The `:=` guarantees this runs only once. This would be added after `include Makefile.venv`.

In this case, using `$(VENV)/entrypoint` is unnecessary, right? The correct tool is always picked up.

Using `make bash` also duplicates the PATH in that case, since the venv is activated, but the PATH is already manipulated.

I like keeping things explicit with the `$(VENV)` prefix.

> Also, be aware that using `$(VENV)` as a dependency in recipes is not the same as using `venv`, and is not recommended. You lose auto-updating capabilities that way - if your requirements.txt changes `$(VENV)` dependency will not trigger an update of the existing virtual environment but `venv` will.

Yeah, that is mentioned in the docs. This was merely a redaction gone wrong on my part, I'm using `venv`.

**Comment by sio** [comment]
> This would be added after include Makefile.venv.

Correct. 

> In this case, using $(VENV)/entrypoint is unnecessary, right? The correct tool is always picked up.

Yes, it's not required after you modify $PATH but it won't lead to any errors if you continue using `$(VENV)/entrypoints`

> Using make bash also duplicates the PATH in that case

Yes, but this is just aesthetics. Operational overhead is almost nonexistent.

> I like keeping things explicit with the $(VENV) prefix.

Me too. I did not take the $PATH modification approach when I first started this project because it's impossible to decide where to put extra $PATH elements for a generic use case. For many projects $VENV should be the first element but it's easy to come up with scenarios where $VENV is better placed to the end or even in the middle of the $PATH.
